README.md
configured-etcd-client
==========================
A small wrapper around etcd to allow configuration from confit.

changelog.md
# Changelog

## 2.0.0
- Update a major version of etcd3 dependency
- Node < 10 is no longer supported
- Fixed a bug where memoize was calling the passed function twice

tests/test_event.js
import tap from 'tap';
import pino from 'pino';
import EtcdClient from '../src/index';

const logger = pino();
const context = { logger, gb: { logger } };

tap.test('test_get_event', async (t) => {
  const etcd = new EtcdClient(context, {
    hosts: [process.env.ETCD_URL || 'http://localhost:2379'],
  });
  const client = await etcd.start();
  const keyName = 'testkey';

  client.addListener('start', (info) => {
    t.equal(info.key, keyName);
    t.equal(info.method, 'get');
  });
  client.addListener('finish', (info) => {
    t.equal(info.key, keyName);
    t.equal(info.method, 'get');
  });

  await client.get(context, keyName);
});

tap.test('test_put_event', async (t) => {
  const etcd = new EtcdClient(context, {
    hosts: [process.env.ETCD_URL || 'http://localhost:2379'],
  });
  const client = await etcd.start();
  const keyName = 'testkey';

  client.addListener('start', (info) => {
    t.equal(info.key, keyName);
    t.equal(info.method, 'put');
  });

  client.addListener('finish', (info) => {
    t.equal(info.key, keyName);
    t.equal(info.method, 'put');
  });

  await client.put(context, keyName, { a: 1 }, 60);
});


tests/test_memoize.js
import tap from 'tap';
import pino from 'pino';
import { v4 as uuidv4 } from 'uuid';
import EtcdClient from '../src/index';

const logger = pino();
const context = { logger, gb: { logger } };

tap.test('test_memoize', async (t) => {
  context.gb = context;
  const etcd = new EtcdClient(context, {
    hosts: [process.env.ETCD_URL || 'http://localhost:2379'],
  });
  const client = await etcd.start();

  const key = `cachekey-${uuidv4()}`;
  const oldValue = `oldval-${uuidv4()}`;
  const newValue = `newval-${uuidv4()}`;
  let value = oldValue;

  const memoFunc = () => value;

  let result = await client.memoize(context, key, memoFunc);
  t.strictEquals(memoFunc(), oldValue);
  t.strictEquals(result, oldValue);

  value = newValue;

  result = await client.memoize(context, key, memoFunc);
  t.strictEquals(memoFunc(), newValue);
  t.strictEquals(result, oldValue);

  const key2 = `concur-${uuidv4()}`;
  let secondRunValue = 'did not wait';
  const delayFunc = async () => {
    await new Promise(accept => setTimeout(accept, 1000));
    secondRunValue = 'new value';
    return 'old value';
  };

  const delayFunc2 = async () => {
    await new Promise(accept => setTimeout(accept, 1000));
    return secondRunValue;
  };

  const memo1 = client.memoize(context, key2, delayFunc, 0);
  await new Promise(accept => setTimeout(accept, 100));
  const memo2 = client.memoize(context, key2, delayFunc2, 0);
  t.strictEquals(await memo1, 'old value', 'Should get the old value the first time');
  t.strictEquals(await memo2, 'new value', 'Should get the new value the second time');

  const key3 = `clear-${uuidv4()}`;
  await client.memoize(context, key3, () => 42, 30);
  t.strictEquals(await client.memoize(context, key3, () => 0, 30), 42, 'Should get the original value');
  await client.clearMemoizedValue(context, key3);
  t.strictEquals(await client.memoize(context, key3, () => 1, 30), 1, 'Should get the new value');
});


tests/test_connection.js
import tap from 'tap';
import pino from 'pino';
import EtcdClient from '../src/index';

const logger = pino();
const context = { logger, gb: { logger } };

tap.test('test_connection', async (connTests) => {
  const etcd = new EtcdClient({ logger }, {
    hosts: [process.env.ETCD_URL || 'http://localhost:2379'],
  });
  const client = await etcd.start();

  const testKey = `test-key-${Date.now()}`;
  const lockKey = `lock-key-${Date.now()}`;

  connTests.test('basic CRUD', async (t) => {
    const objValue = { a: true, b: 3, c: 'four', d: [1, 2, 3] };
    t.notOk(await client.get(context, testKey), 'Should start with no value');
    await client.put(context, testKey, 'helloworld', 1);
    t.strictEquals(await client.get(context, testKey), 'helloworld', 'Should get a key');

    await client.put(context, testKey, objValue, 1);
    t.same(await client.get(context, testKey), objValue, 'Should get a key with object value');
    await new Promise(accept => setTimeout(accept, 2500));
    const afterExpiration = await client.get(context, testKey);
    t.notOk(afterExpiration, 'TTL should work');

    await client.put(context, testKey, 'helloworld');
    await client.del(context, testKey);
    t.notOk(await client.get(context, testKey), 'Del should work');
  });

  connTests.test('recursive get', async (t) => {
    const folder = `folder-${Date.now()}`;
    await client.put(context, `${folder}/${testKey}`, 'helloworld');
    await client.put(context, `${folder}/${lockKey}`, 'i am here');
    const keys = await client.get(context, folder, { recursive: true });
    t.strictEquals(keys[testKey], 'helloworld', 'recursive key/value should match');
    t.strictEquals(keys[lockKey], 'i am here', 'recursive key/value should match');
    t.strictEquals(Object.values(keys).length, 2, 'Should only get two values');
  });

  connTests.test('locking', async (t) => {
    const lock1 = await client.acquireLock(context, lockKey);
    let lock2;
    const lockPromise = client.acquireLock(context, lockKey).then((l) => { lock2 = l; });
    await new Promise(accept => setTimeout(accept, 100));
    t.notOk(lock2, 'After acquiring lock1, lock2 should not fulfill');
    await etcd.releaseLock(lock1);
    await lockPromise;
    t.ok(lock2, 'After releasing lock1, lock2 should fulfill');
    await etcd.releaseLock(lock2);
  });

  connTests.test('watchers', async (t) => {
    const watch = await client.watcher(context, testKey);
    await new Promise((accept) => {
      watch.once('put', (res) => {
        t.strictEquals(res.value.toString(), '"value is changed"', 'New value should match');
        accept();
      });
      client.put(context, testKey, 'value is changed', 1);
    });
    await new Promise((accept) => {
      watch.once('put', (res) => {
        t.strictEquals(res.value.toString(), '"value is so changed"', 'Changed value should match');
        accept();
      });
      client.put(context, testKey, 'value is so changed', 1);
    });
    await watch.cancel();
  });
});


src/index.js
import assert from 'assert';
import { Etcd3, EtcdLockFailedError } from 'etcd3';
import { EventEmitter } from 'events';

function statusCode(error) {
  if (error) {
    return error.errorCode || 'unknown';
  }
  return 0;
}

const LOGGER = Symbol('Logger property for locks');

// This is a bit of a throwback to the etcd2 data model, and it only supports one "level" deep,
// but that's useful enough to keep for us...
function unpackJson(node, prefix = '') {
  return Object.entries(node).reduce((acc, [key, value]) => {
    const keyPart = key.substring(prefix.length).replace(/^\//, '');
    acc[keyPart] = value ? JSON.parse(value) : value;
    return acc;
  }, {});
}

async function delay(ms) {
  return new Promise(accept => setTimeout(accept, ms));
}

async function executeUpToTime(func, maxTimeMs) {
  if (maxTimeMs <= 0) {
    return func();
  }
  const usefulError = new Error('Function timed out');
  return Promise.race([func(), new Promise((accept, reject) => setTimeout(() => reject(usefulError), maxTimeMs))]);
}

export default class Etcd3Client extends EventEmitter {
  constructor(context, opts) {
    super();
    const { hosts, options } = (opts || {});
    this.baseLogger = context.logger || context.gb?.logger;
    assert(this.baseLogger?.info, 'Constructor must have a logger property');
    context.logger.info('Initializing etcd client', {
      hosts: hosts || '<default>',
    });
    const finalOptions = {
      hosts,
      dialTimeout: 1500,
      ...options,
    };
    if (typeof hosts === 'string') {
      finalOptions.hosts = hosts.split(',');
    }
    this.maxRetries = (options && 'maxRetries' in options) ? options.maxRetries : 2;

    this.etcd = new Etcd3(finalOptions);
  }

  async start() {
    return this;
  }

  stop() {
    this.etcd.close();
  }

  finishCall(callInfo, status) {
    this.emit('finish', { status, ...callInfo });
  }

  async get(context, key, options) {
    const callInfo = {
      client: this,
      context,
      key,
      method: 'get',
    };
    const logger = context.gb?.logger || this.baseLogger;
    logger.info('etcd get', { key });
    this.emit('start', callInfo);

    try {
      const value = await (options?.recursive ? this.etcd.getAll().prefix(key) : this.etcd.get(key));
      this.finishCall(callInfo, 0);
      if (!value) {
        logger.info('etcd get empty', { key });
        return value;
      }
      logger.info('etcd get ok', { key });
      return options?.recursive ? unpackJson(value, key) : JSON.parse(value);
    } catch (error) {
      const code = statusCode(error);
      logger.info('etcd get failed', { key, code });
      this.finishCall(callInfo, code);
      throw error;
    }
  }

  /**
   * ttl is in seconds
   */
  async put(context, key, value, ttl) {
    const callInfo = {
      client: this,
      context,
      key,
      value,
      ttl,
      method: 'put',
    };
    const logger = context.gb?.logger || this.baseLogger;
    logger.info('etcd put', { key, ttl });
    this.emit('start', callInfo);

    const stringValue = JSON.stringify(value);
    let lease;
    try {
      if (ttl) {
        lease = this.etcd.lease(ttl);
        await lease.put(key).value(stringValue);
      } else {
        await this.etcd.put(key).value(stringValue);
      }
      this.finishCall(callInfo, 0);
      logger.info('etcd was put', { key });
    } catch (error) {
      const code = statusCode(error);
      logger.info('etcd put failed', { key, code });
      this.finishCall(callInfo, code);
      throw error;
    } finally {
      if (lease) {
        lease.release();
      }
    }
  }

  async del(context, key) {
    const callInfo = {
      client: this,
      context,
      key,
      method: 'del',
    };
    const logger = context.gb?.logger || this.baseLogger;
    logger.info('etcd del', { key });
    this.emit('start', callInfo);

    try {
      await this.etcd.delete().key(key);
      this.finishCall(callInfo, 0);
      logger.info('etcd del ok', { key });
    } catch (error) {
      const code = statusCode(error);
      logger.info('etcd del failed', { key, code });
      this.finishCall(callInfo, code);
      throw error;
    }
  }

  async acquireLock(context, key, timeout = 10, maxWait = 30) {
    const callInfo = {
      client: this,
      context,
      key,
      method: 'acquireLock',
    };
    const logger = context.gb?.logger || this.baseLogger;
    logger.info('etcd acquire', { key });
    this.emit('start', callInfo);

    const lock = this.etcd.lock(key).ttl(timeout);
    lock[LOGGER] = { logger, key };

    const startTime = Date.now();
    let attempt = 0;
    while (Date.now() - startTime < maxWait * 1000) {
      attempt += 1;
      try {
        // eslint-disable-next-line no-await-in-loop
        await lock.acquire();
        const waitTime = Date.now() - startTime;
        logger.info('Acquired lock', { key, waitTime });
        this.finishCall(callInfo, 'acq');
        return lock;
      } catch (error) {
        logger.warn('Lock contention', { key, attempt });
        if (!(error instanceof EtcdLockFailedError)) {
          this.finishCall(callInfo, 'err');
          throw error;
        }
        // eslint-disable-next-line no-await-in-loop
        await delay(250 * attempt);
      }
    }
    this.finishCall(callInfo, 'timeout');
    const waitTime = Date.now() - startTime;
    const error = new Error('Timed out waiting for lock');
    error.waitTime = waitTime;
    throw error;
  }

  // eslint-disable-next-line class-methods-use-this
  async releaseLock(lock) {
    try {
      const { key, logger } = lock[LOGGER] || {};
      await lock.release();
      if (logger) {
        logger.info('Lock released', { key });
        delete lock[LOGGER];
      }
    } catch (error) {
      // Nothing to do for this error - eat it
    }
  }

  /**
   * This method is expensive. Please don't call it unless you need it.
   * For example: when making a critical area idempotent.
   * Even if you think you need it consult #guild-server-devs first.
   */
  async memoize(context, key, func, ttl = 60 * 5, timeout = 10, maxWait = 30, maxExecutionTime = 0) {
    const callInfo = {
      client: this,
      context,
      key,
      method: 'memoize',
    };

    const logger = context.gb?.logger || this.baseLogger;
    logger.info('etcd memoize', { key });
    this.emit('start', callInfo);

    let lock;
    let value;
    let fnError;
    const lockKey = `${key}-lock`;
    const valueKey = `${key}-value`;

    try {
      value = await this.get(context, valueKey);
      if (value) {
        this.finishCall(callInfo, 'val-prelock');
      } else {
        lock = await this.acquireLock(context, lockKey, timeout, maxWait);
        value = await this.get(context, valueKey);
        if (value) {
          this.finishCall(callInfo, 'val-postlock');
        } else {
          logger.info('etcd memoize exec', { key });
          try {
            value = await executeUpToTime(func, maxExecutionTime * 1000);
          } catch (error) {
            fnError = true;
            throw error;
          }
          if (ttl !== 0) {
            await this.put(context, valueKey, value, ttl);
          }
          this.finishCall(callInfo, 'val-eval');
        }
      }
    } catch (error) {
      logger.error(fnError ? 'etcd memoize fn failed' : 'etcd memoize failed', error);
      this.finishCall(callInfo, fnError ? 'fn-error' : 'error');
      throw error;
    } finally {
      if (lock) {
        this.releaseLock(lock);
      }
    }

    return value;
  }

  async clearMemoizedValue(context, key) {
    return this.del(context, `${key}-value`);
  }

  async watcher(context, key) {
    const logger = context.gb?.logger || this.baseLogger;
    logger.info('Starting etcd watch', { key });

    return this.etcd.watch()
      .key(key)
      .create()
      .then((watcher) => {
        watcher
          .on('disconnected', () => logger.info('etcd watcher disconnected', { key }))
          .on('error', error => logger.warn('etcd watcher error', error))
          .on('connected', () => logger.info('etcd watcher connected', { key }));
        return watcher;
      });
  }
}
